tsl/src/remote/copy_fetcher.c

```diff
@@ -240,6 +241,66 @@ copy_data_check_header(StringInfo copy_data)
 	}
 }
 
+/*
+ * End the COPY after receiving EOF or canceling a query (e.g., due to a LIMIT
+ * being reached).
+ *
+ * This should be called after canceling a query, or, after reading all data,
+ * the file trailer, and getting an EOF return value.
+ */
+static void
+end_copy(CopyFetcher *fetcher)
+{
+	PGconn *conn = remote_connection_get_pg_conn(fetcher->state.conn);
+	PGresult *final_pgres = NULL;
+	PGresult *pgres = NULL;
+	/* Expect COMMAND OK if we received EOF. If the query was canceled (e.g.,
+	 * due to a LIMIT), expect FATAL_ERROR (query failed) */
+	ExecStatusType expected_status = fetcher->state.eof ? PGRES_COMMAND_OK : PGRES_FATAL_ERROR;
+	ExecStatusType received_status;
+
+	Assert(fetcher->state.open);
+
+	/* Read results until NULL */
+	while ((pgres = PQgetResult(conn)))
+	{
+		if (final_pgres == NULL)
+			final_pgres = pgres;
+		else
+			remote_result_close(pgres);
+	}
+
+	received_status = PQresultStatus(final_pgres);
+	remote_result_close(final_pgres);
+
+	if (received_status != expected_status)
+		remote_connection_elog(fetcher->state.conn, ERROR);
+
+	fetcher->state.open = false;
+	remote_connection_set_status(fetcher->state.conn, CONN_IDLE);
+}
+
+/*
+ * Prematurely end the COPY before EOF is received.
+ *
+ * This happens for queries that are abandoned before reaching EOF, e.g., when
+ * using LIMITs.
+ */
+static void
+end_copy_before_eof(CopyFetcher *fetcher)
+{
+	/*
+	 * The fetcher state might not be open if the fetcher got initialized but
+	 * never executed due to executor constraints.
+	 */
+	if (!fetcher->state.open)
+		return;
+
+	Assert(!fetcher->state.eof);
```

Comment:
Probably also doesn't make sense to cancel if we have received the file trailer. I'm just going through all usages of `trailer`/`eof` to check if they're used in the same way.